BUG: fix DataFrame.apply returning wrong result when dealing with dtype (#28773)

gorski-m opened the pull request:
- [x] closes #28773
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

The DataFrame.apply was sometimes returning wrong result when we passed
function, that was dealing with dtypes. It was caused by retrieving
the DataFrame.values of whole DataFrame, and applying the function
to it: values are represented by NumPy array, which has one
type for all data inside. It sometimes caused treating objects
in DataFrame as if they had one common type. What's worth mentioning,
the problem only existed, when we were applying function on columns.

The implemented solution "cuts" the DataFrame by columns and applies
function to each part, as it was whole DataFrame. After that, all
results are concatenated into final result on whole DataFrame.
The "cuts" are done in following way: the first column is taken, and
then we iterate through next columns and take them into first cut
while their dtype is identical as in the first column. The process
is then repeated for the rest of DataFrame.

Commit:
BUG: fix DataFrame.apply returning wrong result when dealing with dtype (#28773)

The DataFrame.apply was sometimes returning wrong result when we passed
function, that was dealing with dtypes. It was caused by retrieving
the DataFrame.values of whole DataFrame, and applying the function
to it: values are represented by NumPy array, which has one
type for all data inside. It sometimes caused treating objects
in DataFrame as if they had one common type. What's worth mentioning,
the problem only existed, when we were applying function on columns.

The implemented solution "cuts" the DataFrame by columns and applies
function to each part, as it was whole DataFrame. After that, all
results are concatenated into final result on whole DataFrame.
The "cuts" are done in following way: the first column is taken, and
then we iterate through next columns and take them into first cut
while their dtype is identical as in the first column. The process
is then repeated for the rest of DataFrame

Commit:
Merge remote-tracking branch 'upstream/master' into apply_bug

pep8speaks commented:
Hello @Reksbril! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





























* In the file [`pandas/tests/arrays/sparse/test_array.py`](https://github.com/pandas-dev/pandas/blob/e2161648c409736926e94b9ca6073c25848cb23b/pandas/tests/arrays/sparse/test_array.py):

> [Line 663:24](https://github.com/pandas-dev/pandas/blob/e2161648c409736926e94b9ca6073c25848cb23b/pandas/tests/arrays/sparse/test_array.py#L663): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 668:24](https://github.com/pandas-dev/pandas/blob/e2161648c409736926e94b9ca6073c25848cb23b/pandas/tests/arrays/sparse/test_array.py#L668): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','







* In the file [`pandas/tests/indexing/test_callable.py`](https://github.com/pandas-dev/pandas/blob/e2161648c409736926e94b9ca6073c25848cb23b/pandas/tests/indexing/test_callable.py):

> [Line 20:39](https://github.com/pandas-dev/pandas/blob/e2161648c409736926e94b9ca6073c25848cb23b/pandas/tests/indexing/test_callable.py#L20): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 23:39](https://github.com/pandas-dev/pandas/blob/e2161648c409736926e94b9ca6073c25848cb23b/pandas/tests/indexing/test_callable.py#L23): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 93:42](https://github.com/pandas-dev/pandas/blob/e2161648c409736926e94b9ca6073c25848cb23b/pandas/tests/indexing/test_callable.py#L93): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','

jbrockmendel commented:
>  The rest of changes are result of black pandas reformatting.

I guess you're using a different version of black than our CI is?  Regardless, can you revert non-relevant changes so we can focus on the important parts

gorski-m committed:
Reformatting

Reksbril commented:
@jbrockmendel done.

jbrockmendel commented:
This is a lot of new code.  A couple things:

- Assuming this is the correct solution to this problem, the new code would belong in core.apply
- the "cuts" are a lot like the `DataFrame._data.blocks` that already exist.  Usually we discourage people from bothering with blocks but it is an option
- The hard part of the problem is identifying when you should be operating column-wise (i.e. not call .values).  Finding an elegant-ish way of doing that will be key here.

gorski-m committed:
Resolve conflict with iterator variables

Reksbril commented:
@jreback 
I've resolved a minor problem which caused most of tests to fail.

>     * the "cuts" are a lot like the `DataFrame._data.blocks` that already exist.  Usually we discourage people from bothering with blocks but it is an option
What's the benefit from using `DataFrame._data.blocks` instead of current solution? I'm just using DataFrame.iloc to get only part of DF and it seems to be very "clean" approach.
>     * The hard part of the problem is identifying when you should be operating column-wise (i.e. not call .values).  Finding an elegant-ish way of doing that will be key here.
I'm not sure if we can determine that at all. The obvious thing we can check is if all dtypes are the same (I think it's already done in current code and some specific apply variant is used in that case). As the functions given in input can be very complicated, the only solution to this problem I see, is to give some experimental input to the function, to determine if it relies on dtypes. However, that would be neither elegant nor easy to do (if even possible).

jbrockmendel commented:
@Reksbril you're going to have to go through `frame_apply` and try to adapt the existing machinery rather than pile a new layer on top of it.  Within FrameApply there is a `apply_series_generator` that exists precisely to have a function operate column-by-column.  So a good solution to this problem would be to edit FrameApply.apply_standard so that this case goes through the apply_series_generator route.

jreback commented on pandas/core/frame.py:
```diff
@@ -6564,16 +6564,63 @@ def apply(self, func, axis=0, raw=False, result_type=None, args=(), **kwds):
-            kwds=kwds,
-        )
-        return op.get_result()
+        # Old apply function, which will be used for each part of DataFrame
```
*all* of this should be done in pandas.core.apply

jreback reviewed (changes requested):
pls delve deeper in how apply; we already have a column-by-column approach, your test case needs to inform when to use that

Commit:
Change approach to #28773 fix

In new solution, existing machinery is used to apply the function
column-wise, and to recreate final result.

gorski-m committed:
Merge branch 'master' into apply_bug

gorski-m committed:
Update apply.py

Reksbril commented:
@jreback ping

Reksbril commented:
@jreback ping

jbrockmendel commented:
@Reksbril needs rebase, ill take a look

jbrockmendel commented on pandas/core/apply.py:
```diff
@@ -274,13 +274,16 @@ def apply_standard(self):
             #  cannot handle MultiIndex
             and not isinstance(self.agg_axis, ABCMultiIndex)
-        ):
+        )
```
what happens if you change this condition to also rule out mixed-dtypes?  that would prevent the call to self.values which is what does the object-dtype casting that is causing the problem

Reksbril replied:
Done. I forgot to add this additional rule before

Commit:
Filter out mixed-dtypes from regular standard_apply procedure

gorski-m committed:
Merge branch 'master' into apply_bug

gorski-m committed:
Remove whitespace from test_apply.py

jreback commented on pandas/core/apply.py:
```diff
@@ -274,13 +274,17 @@ def apply_standard(self):
             #  cannot handle MultiIndex
             and not isinstance(self.agg_axis, ABCMultiIndex)
-        ):
+            and len(set(self.dtypes))
```
use ```self.dtypes.nunique > 0```

call this something besides ``flag``, maybe ``can_reduce``

jreback commented on pandas/core/apply.py:
```diff
@@ -274,13 +274,17 @@ def apply_standard(self):
-        ):
+            and len(set(self.dtypes))
+        )
+        return_result = None
```
why is this needed?

Reksbril replied:
I'm using it to determine if line 314 was run.

jreback commented on pandas/core/apply.py:
```diff
@@ -308,11 +312,19 @@ def apply_standard(self):
                 pass
             else:
-                return self.obj._constructor_sliced(result, index=labels)
+                return_result = self.obj._constructor_sliced(result, index=labels)
```
what are you trying to do here? we have a specific subclass for Series where this could live, but I am not clear what you are trying to catch

jreback commented on pandas/core/apply.py:
```diff
@@ -308,11 +312,19 @@ def apply_standard(self):
         # compute the result using the series generator
         results, res_index = self.apply_series_generator()
 
+        if flag and return_result is not None:
```
what is this case tryng to catch?

Reksbril replied:
There are 2 cases
1) can_reduce is False - the function doesn't change its behaviour, compared to previous version
2) can_reduce is True - the return_result is calculated like before, but it's returned only if we don't use apply on columns or we don't have mixed dtypes. In other case I use apply_series_generator to obtain column-by-column result and after that I'm wrapping it using _constructor_sliced. It's done this way to preserve the "old" way - it was solution to problem with my function sometimes returning Series instead of DataFrame (or the other way). I assumed, that already existing way of getting result would be better, than thinking about some new solutions.

Reksbril replied:
@jreback

jreback commented on pandas/tests/frame/test_apply.py:
```diff
@@ -691,6 +691,12 @@ def test_apply_dup_names_multi_agg(self):
+        # GH 28773
+        df = DataFrame({"col_1": [1, 2, 3], "col_2": ["hi", "there", "friend"]})
+        expected = Series(data=["int64", "object"], index=["col_1", "col_2"])
+        tm.assert_series_equal(df.apply(lambda x: x.dtype), expected)
```
use 

result=
expected=
assert_.....

jreback reviewed (changes requested):
your patch makes this much more complex; pls try to simplify

Commit:
Refactor

Commit:
Black pandas refactor

Reksbril commented:
@jreback ping

WillAyd commented:
@Reksbril looks like a merge conflict and CI failure - can you fix those up and someone will take a look

jbrockmendel commented on pandas/core/apply.py:
```diff
@@ -274,13 +274,16 @@ def apply_standard(self):
+
+        if can_reduce:
 
             values = self.values
```
@Reksbril the underlying problem here is that this call to `self.values` is casting to object dtype which you want to avoid.  if you add to the check in 277-283 `and self.obj._is_homogeneous_type` that should solve the whole thing

jbrockmendel commented:
@Reksbril can you rebase and address comments

Reksbril commented:
Yeah, I will. I'm sorry for long delay.

Commit:
Merge remote-tracking branch 'upstream/master' into apply_bug

Commit:
apply fix

Reksbril commented:
I won't be able to finish this for a few more months. The current version is the best I could do right now.

simonjayhawkins commented:
> I won't be able to finish this for a few more months. The current version is the best I could do right now.

Thanks @Reksbril for working on this. closing to clear queue for now. ping when you want to continue and will reopen.
